### ISSUE TITLE
The webhook should reject IPPool deletion if any IP has been allocated

### ISSUE THREAD

**Issue by jessehu** [metal3-io/ip-address-manager#135]
**What steps did you take and what happened:**
[A clear and concise description on how to REPRODUCE the bug.]

**What did you expect to happen:**

**Anything else you would like to add:**
Looks like it doesn't handle the delete request: https://github.com/metal3-io/ip-address-manager/blob/main/api/v1alpha1/ippool_webhook.go#L137

**Environment:**

- Cluster-api version:
- CAPM3 version:
- IPAM version:
- Minikube version:
- environment (metal3-dev-env or other):
- Kubernetes version: (use `kubectl version`):

/kind bug

**Comment by furkatgofurov7** [comment]
I am not sure tbh if we want to have logic for ippool webhook deletion.. reconcileDelete() does check for the allocations https://github.com/metal3-io/ip-address-manager/blob/31c03f129143921604c07591d49f32a118200216/controllers/ippool_controller.go#L156-L171 . Also, what is the common way to handle the webhooks, seems CAPI also does handle the webhook the same way https://github.com/kubernetes-sigs/cluster-api/blob/main/api/v1beta1/machine_webhook.go#L88-L90?

/cc @schrej if you have other opinions.

**Comment by schrej** [comment]
Well, there are only two options on how to handle deletion: finalizers and webhooks.

Finalizers are the safe way, as they prevent deletion on the apiserver and even work when the webhook isn't deployed.
Webhooks are the user friendly way, as they properly prevent deletion, and the resource doesn't stay in a deleted state for a long time, as is the case with Finalizers (you can't un-delete a resource).
In CAPI this was discussed in https://github.com/kubernetes-sigs/cluster-api/issues/5498 for example.

Either solution, or even a combination of both, is fine. Personally I'd probably prefer the combined approach, as you don't end up with deleted pools that remain in use for a long time that way.

**Comment by jessehu** [comment]
As you can't un-delete a resource in K8s, the webhook approach can prevent IPPool from going into Deleting state when it's in used. In case it's deleted unintentionally.

**Comment by furkatgofurov7** [comment]
Right now, the deletion is accepted, but we wait until all addresses are released before actually deleting.
Now #137 would move the responsibility onto the user. I am not sure if that is a good thing. The user won't be able to send a delete on the object and just wait until it is deleted. The user will instead have to retry deleting it as long as addresses are allocated.  But then the behavior is a bit more straightforward...

**Comment by jessehu** [comment]
Hi @furkatgofurov7 got your point. When the user got error message that "IPPool is in used so can not be deleted", would he/she release the used IPs in this pool first then delete this pool again?

**Comment by furkatgofurov7** [comment]
> "IPPool is in used so can not be deleted", would he/she release the used IPs in this pool first then delete this pool again?

Yup, that is how it is now when the delete is issued in the objects

**Comment by furkatgofurov7** [comment]
But we are also fine also to have `combined` handling as @schrej suggested.

**Comment by jessehu** [comment]
Thanks @furkatgofurov7. Could you help proceed https://github.com/metal3-io/ip-address-manager/pull/137 ? It's ready for review and CI passed.

**Comment by furkatgofurov7** [comment]
triage/accepted

**Comment by furkatgofurov7** [comment]
prow is bit slow for requests but eventually should be able to label this. Otherwise I will do it manually.

**Comment by Rozzii** [comment]
What is the status of this @furkatgofurov7 ?

**Comment by furkatgofurov7** [comment]
#137 referring as a fix to this, but that seems to have some issues passing the CI?
